ui/common/Hasher.tsx

```diff
@@ -0,0 +1,154 @@
+'use client'
+
+import clsx from 'clsx'
+import { useEffect, useState } from 'react'
+import { Button } from 'shared'
+import sjcl from 'sjcl'
+
+export default function Hasher({
+  answer,
+  label,
+  onChange,
+  auto,
+  input,
+}: {
+  answer: string | number
+  label: string
+  onChange: any
+  auto: boolean
+  input: string
+}) {
+  const [hash, setHash] = useState('')
+  const [clicked, setClicked] = useState(false)
+
+  const handleAutoChange = () => {
+    let randomNum = (Math.random() * 1000000000000000000).toString()
+    setHash(sjcl.codec.hex.fromBits(sjcl.hash.sha256.hash(randomNum)))
+    console.log(randomNum)
+  }
+
+  const handleHash = (input) => {
+    if (input) {
+      const bitArray = sjcl.hash.sha256.hash(input)
+      setHash(sjcl.codec.hex.fromBits(bitArray))
+    } else {
+      setHash('')
+    }
+  }
+
+  function handleClick() {
+    setClicked(!clicked)
+  }
+
+  useEffect(() => {
+    const miner =
+      clicked &&
+      typeof answer === 'string' &&
+      !hash.startsWith(answer) &&
+      setInterval(() => {
+        handleAutoChange()
+      }, 10)
+    !clicked && handleHash(input)
+    if (typeof answer === 'string') {
+      onChange(hash.substring(0, answer.length))
+    } else {
+      onChange(hash)
+    }
+    return () => {
+      clearInterval(miner)
+    }
+  })
+
+  return (
+    <>
+      {!auto && (
+        <>
+          <div className="flex flex-col">
+            <h2 className="text-left text-[18px] font-bold md:text-center">
+              {label}
+            </h2>
+            <p
+              className={clsx(
+                'top-0 left-0 h-full w-4/5 resize-none overflow-hidden break-words bg-transparent text-left font-space-mono text-[18px] leading-[180%] tracking-[1px] outline-none md:w-3/5 md:text-[30px] md:tracking-[5px]',
+                {
+                  'overlay-complete':
+                    (typeof answer === 'string' &&
+                      hash.startsWith(answer) === true) ||
+                    (typeof answer === 'number' && input.length >= answer),
+                  'overlay-incomplete':
+                    (typeof answer === 'string' &&
+                      hash.startsWith(answer) === false) ||
+                    (typeof answer === 'number' && input.length < answer),
+                  underscore: !hash || !answer,
+                }
+              )}
+            >
+              {hash
+                ? `${hash.substring(0, 4)} ${hash.substring(
```

Comment:
I'm guessing this code is used to `splitastringlikethis` into `spli tast ring like this`, I don't think this is the way to do that. You should use a loop or something. I had a similar challenge [here](https://github.com/saving-satoshi/saving-satoshi/blob/master/ui/lesson/ScriptingChallenge/Runner/Hasher.tsx#L36), perhaps you can extract that to a helper function that we can reuse in this challenge and in the scripting challenge.

Reply:
ATM I have a small regex function that replaces that long mess. But I like your idea of implementing your helper function into a separate function to reuse